src/gauges.rs

```diff
@@ -0,0 +1,250 @@
+            .set(epoch_info.transaction_count.unwrap_or_default() as i64);
+        self.slot_height.set(epoch_info.absolute_slot as i64);
+        self.current_epoch.set(epoch_info.epoch as i64);
+        self.current_epoch_first_slot.set(first_slot);
+        self.current_epoch_last_slot.set(last_slot);
+
+        Ok(())
+    }
+
+    // For now, this will export the IP addresses of active voting accounts with a node.
+    // TODO: This needs to actually export to a Prometheus gauge.
+    pub async fn export_ip_addresses(
+        &self,
+        nodes: &[RpcContactInfo],
+        vote_accounts: &RpcVoteAccountStatus,
+        api_key: &MaxMindAPIKey,
+        cache: &GeoCache,
+    ) -> anyhow::Result<()> {
+        // Define all types here
+        type RpcInfo = (RpcContactInfo, RpcVoteAccountInfo);
+        type RpcInfoMaybeGeo = (RpcContactInfo, RpcVoteAccountInfo, Option<CityApiResponse>);
+        type RpcInfoGeo = (RpcContactInfo, RpcVoteAccountInfo, CityApiResponse);
+
+        // All nodes that are validators
+        let validator_nodes = {
+            // Mapping of pubkey -> vote account info.
+            // Convert into a mapping to avoid O(n) searches
+            let pubkeys = vote_accounts
+                .current
+                .iter()
+                .cloned()
+                .map(|vote| (vote.node_pubkey.to_string(), vote))
+                .collect::<HashMap<String, RpcVoteAccountInfo>>();
+
+            nodes
+                .iter()
+                .cloned()
+                .filter_map(|contact| {
+                    pubkeys
+                        .get(&contact.pubkey)
+                        .map(|vote| (contact, vote.clone()))
+                })
+                .collect::<Vec<RpcInfo>>()
+        };
+
+        // Separate cached from uncached
+        let (cached, uncached): (Vec<RpcInfoMaybeGeo>, Vec<RpcInfoMaybeGeo>) = validator_nodes
+            .into_iter()
+            .map(|(contact, vote)| {
+                let cached = cache
+                    .fetch_ip_address_with_invalidation(
+                        &get_rpc_contact_ip(&contact).with_context(|| {
+                            format!("Validator node has no IP: {:?} {:?}", contact, vote)
+                        })?,
+                        |date| date + Duration::week() < OffsetDateTime::now_utc().date(),
+                    )?
+                    .map(|geo| geo.response);
+                Ok((contact, vote, cached))
+            })
+            .collect::<anyhow::Result<Vec<RpcInfoMaybeGeo>>>()?
+            .into_iter()
+            .partition(|(_, _, db)| db.is_some());
+
+        let mut geolocations = cached
+            .into_iter()
+            // This unwrap is safe because we have already partitioned into Some and Nones.
+            .map(|(c, v, db)| (c, v, db.unwrap()))
+            .collect::<Vec<RpcInfoGeo>>();
+
+        // For uncached, request them from maxmind.
+        let client = reqwest::Client::new();
```

Comment:
Can this client be part of the `PrometheusGauges` state so that you don't have to create one every time?

Reply:
Perhaps we could move it into a `Lazy` static, as [according to the docs](https://docs.rs/reqwest/0.11.3/reqwest/struct.Client.html)...
> The `Client` holds a connection pool internally, so it is advised that you create one and reuse it.
> You do not have to wrap the `Client` it in an `Rc` or `Arc `to reuse it, because it already uses an `Arc` internally.

Reply:
You could just fire it up in `PrometheusGauges::default` by `let client = reqwest::Client::new();`
No need for laziness IMO.

Reply:
As a side note, the `default` method is rather specialised. Maybe it's better to move it to `PrometheusGauges::new`.

Reply:
Fixed in 794584d